---
title: "Maaslin2 on 16S and ITS with confounders Genus detailed result"
output: html_document
date: "2023-10-17"
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## R Markdown
```{r}
rm(list = ls())
```

```{r}
library(here)
library(tidyverse)
library(microbiome)
library(microbiomeMarker)
library(phyloseq)
library(pheatmap)
library(RColorBrewer)
# Load the package
library(ComplexHeatmap)
library(circlize)
library(Maaslin2)
library(ggbreak)
library(patchwork)
```

```{r}
```

### Import data

```{r}
Plaque538_16S_AllOTU_5perc_ps =   readRDS(here("../RP_240306_Plaque538_samples/Plaque538_Combined/Plaque538_16S/Plaque538_16S_A2_pseq_objects",
                          "Plaque538_16S_AllOTU_5perc_ps_17_vars_240306.Rds"))
Plaque538_16S_AllOTU_5perc_ps

## This is a good to remove the bad_empty names eg. "Unknown" also, if required
## the aggregate_taxa only keeps the names at a particular tax-level eg Genus with prerix "g__"
Plaque538_16S_Genus_ps = 
  microbiomeMarker::aggregate_taxa( Plaque538_16S_AllOTU_5perc_ps, level = "Genus", verbose = FALSE) %>%
  phyloseq::tax_glom(
  .,
  taxrank = "Genus",
  bad_empty = c(NA, "", " ", "\t", "__") ) 
Plaque538_16S_Genus_ps
```


# function to modify the taxa names

```{r}
# Your custom function to replace characters
replace_chars <- function(names) {
  gsub("\\(|\\)|\\[|\\]|-|&", "", names)
}
```

## modify the taxa names

```{r}
# Apply replace_chars to the row names of the OTU table
modified_row_names <- replace_chars(rownames(otu_table(Plaque538_16S_Genus_ps)))

# Assign back to the phyloseq object
taxa_names(Plaque538_16S_Genus_ps) <- modified_row_names
rownames(otu_table(Plaque538_16S_Genus_ps)) <- modified_row_names


# Modify taxa names
new_taxa_names_16S_Genus <- gsub("g__", "B_g_", taxa_names(Plaque538_16S_Genus_ps))

# Apply new names back to the phyloseq object
taxa_names(Plaque538_16S_Genus_ps)          <- new_taxa_names_16S_Genus
rownames(otu_table(Plaque538_16S_Genus_ps)) <- new_taxa_names_16S_Genus


# Confirm the changes
Plaque538_16S_Genus_ps %>% otu_table() %>% rownames() %>% head()

```


### Normalize in Microbiomemarker

```{r}
## AFter filtering some of the samples have all OTUs 0, hence it is important to add one to mitigate this problem
Plaque538_16S_Genus_ps_CPM =
  phyloseq::prune_samples(sample_sums(Plaque538_16S_Genus_ps) != 0, Plaque538_16S_Genus_ps) %>% 
  microbiomeMarker::normalize (., "CPM")

Plaque538_16S_Genus_ps_CPM
```

### Import data

```{r}
Plaque538_ITS_AllOTU_5perc_ps =   readRDS(here("../RP_240306_Plaque538_samples/Plaque538_Combined/Plaque538_ITS/Plaque538_ITS_A2_pseq_objects",
                          "Plaque538_ITS_AllOTU_5perc_ps_17_vars_240306.Rds"))
Plaque538_ITS_AllOTU_5perc_ps

## This is a good to remove the bad_empty names eg. "Unknown" also, if required
## the aggregate_taxa only keeps the names at a particular tax-level eg Genus with prerix "g__"
Plaque538_ITS_Genus_ps = 
  microbiomeMarker::aggregate_taxa( Plaque538_ITS_AllOTU_5perc_ps, level = "Genus", verbose = FALSE) %>%
  phyloseq::tax_glom(
  .,
  taxrank = "Genus",
  bad_empty = c(NA, "", " ", "\t", "__") ) 
Plaque538_ITS_Genus_ps
```


## modify the taxa names

```{r}
# Apply replace_chars to the row names of the OTU table
modified_row_names <- replace_chars(rownames(otu_table(Plaque538_ITS_Genus_ps)))

# Assign back to the phyloseq object
taxa_names(Plaque538_ITS_Genus_ps) <- modified_row_names
rownames(otu_table(Plaque538_ITS_Genus_ps)) <- modified_row_names


# Modify taxa names
new_taxa_names_ITS_Genus <- gsub("g__", "F_g_", taxa_names(Plaque538_ITS_Genus_ps))

# Apply new names back to the phyloseq object
taxa_names(Plaque538_ITS_Genus_ps)          <- new_taxa_names_ITS_Genus
rownames(otu_table(Plaque538_ITS_Genus_ps)) <- new_taxa_names_ITS_Genus


# Confirm the changes
Plaque538_ITS_Genus_ps %>% otu_table() %>% rownames() %>% head()

```


### Normalize in Microbiomemarker

```{r}
## AFter filtering some of the samples have all OTUs 0, hence it is important to add one to mitigate this problem
Plaque538_ITS_Genus_ps_CPM =
  phyloseq::prune_samples(sample_sums(Plaque538_ITS_Genus_ps) != 0, Plaque538_ITS_Genus_ps) %>% 
  microbiomeMarker::normalize (., "CPM")

Plaque538_ITS_Genus_ps_CPM
```
## Remove genus with Incertae_sedis


```{r}
# Extract the taxonomy table
taxonomy_table <- tax_table(Plaque538_ITS_Genus_ps_CPM)

# Identify taxa where "gen_Incertae_sedis" appears in any part of the Genus column (or other relevant columns)
# Use grepl for partial matches
incertae_sedis_taxa <- rownames(taxonomy_table)[grepl("gen_Incertae_sedis", taxonomy_table[, "Genus"])]

# Remove taxa with "gen_Incertae_sedis" from the phyloseq object
Plaque538_ITS_Genus_ps_CPM <- prune_taxa(!(taxa_names(Plaque538_ITS_Genus_ps_CPM) %in% incertae_sedis_taxa), Plaque538_ITS_Genus_ps_CPM)

Plaque538_ITS_Genus_ps_CPM
```


## Remove the variables not required from the sample data
```{r}
# remove the variables not required
variables_to_remove <- c("Child_overall_health",
                         "Material_depr_score",
                         "Vaginal_birth",
                         "Breastfeed_child",
                         "Bottlefeed_child",
                         "Bedtime_snack",
                         "Child_dental_health",
                         "Times_day_brushed",
                         "Fluoride_toothpaste",
                         "Social_depr_score",
                         "ECOHIS_total_score",
                         "NutriSTEP_total_score"
                         )


Plaque538_16S_Genus_ps_CPM@sam_data <- 
  Plaque538_16S_Genus_ps_CPM@sam_data[, !(colnames(Plaque538_16S_Genus_ps_CPM@sam_data) %in% variables_to_remove)]
Plaque538_16S_Genus_ps_CPM

Plaque538_ITS_Genus_ps_CPM@sam_data <- 
  Plaque538_ITS_Genus_ps_CPM@sam_data[, !(colnames(Plaque538_ITS_Genus_ps_CPM@sam_data) %in% variables_to_remove)]
Plaque538_ITS_Genus_ps_CPM

```

## Merge the two phyloseq objects
```{r}
# Plaque538_16S_ITS_Genus_ps = merge_phyloseq(Plaque538_16S_Genus_ps_CPM,Plaque538_ITS_Genus_ps_CPM)#,Plaque538_16S_KEGG_ps)
# Plaque538_16S_ITS_Genus_ps
```

We will apply Maaslin2 separately on 16S and ITS data and later combine the output to get a single heatmap

```{r}
# confounders = c()
confounders = c("Sex", "Age", "Urban_status", "SEFI_score")
```


## Maaslin2 output directory

```{r}
Maaslin2_Genus_output_dir_16S = here("Samples538_A_Geno_Microbiome/A2_Maaslin",
                           "Maaslin2_16S_Genus_output")
# if file error quit sesssion and connect

if (!dir.exists(Maaslin2_Genus_output_dir_16S)) {
  dir.create(Maaslin2_Genus_output_dir_16S, recursive = TRUE)
}

# Do it for its data
Maaslin2_Genus_output_dir_ITS = here("Samples538_A_Geno_Microbiome/A2_Maaslin",
                           "Maaslin2_ITS_Genus_output")
# if file error quit sesssion and connect

if (!dir.exists(Maaslin2_Genus_output_dir_ITS)) {
  dir.create(Maaslin2_Genus_output_dir_ITS, recursive = TRUE)
}

```


```{r message=FALSE}
# Initialize lists to temporarily hold the data
list_coef_16S <- list()
list_qval_16S <- list()

pseq = Plaque538_16S_Genus_ps_CPM

minimum_prevalence = 0.05


# Assuming pseq is your data frame or a similar object
col_without_confounders = setdiff(colnames(sample_data(pseq)), confounders)

# Iterate through each variable to run Maaslin2 and filter results
for(var in col_without_confounders) {
#for(var in c("ECC_status", "Child_dental_health")) { # for try
  # Run Maaslin2 analysis
  Maaslin2_res_Genus_16S <- Maaslin2(input_data = t(otu_table(pseq)),
         input_metadata = sample_data(pseq) %>% data.frame(),
         output = Maaslin2_Genus_output_dir_16S,
         normalization = "CLR",
         transform = "NONE",
         fixed_effects =  c(var, confounders),
         cores = 40,
         max_significance = 1.1,
         min_prevalence = minimum_prevalence
           )
  
    # Filter and transform the output data frame
  Maaslin2_res_Genus_16S_df <- Maaslin2_res_Genus_16S$results %>% 
                 data.frame() %>% 
                  rename(OTUs = feature ) %>% 
                 filter(metadata  == var) %>% 
                 dplyr::select(OTUs, coef, qval,stderr)
  
  # Rename the coef and qval columns to the current variable
  df_coef_16S <- Maaslin2_res_Genus_16S_df %>% dplyr::select(OTUs, coef) %>% 
                     rename(!!var := coef)
                     
  df_qval_16S <- Maaslin2_res_Genus_16S_df %>% dplyr::select(OTUs, qval) %>% 
                      rename(!!var := qval)
  
  # Store in lists
  list_coef_16S[[var]] <- df_coef_16S
  list_qval_16S[[var]] <- df_qval_16S
}

df_coef_16S <- list_coef_16S[[1]]
# for(i in 2:length(list_coef_16S)) {
#   Genus_coef_16S <- full_join(Genus_coef_16S, list_coef_16S[[i]], by = "OTUs")
# }

df_qval_16S <- list_qval_16S[[1]]
# for(i in 2:length(list_qval_16S)) {
#   Genus_qval_16S <- full_join(Genus_qval_16S, list_qval_16S[[i]], by = "OTUs")
# }
Maaslin2_res_Genus_16S
```

## Maaslin2 on ITS
```{r}
# Initialize lists to temporarily hold the data
list_coef_ITS <- list()
list_qval_ITS <- list()

pseq = Plaque538_ITS_Genus_ps_CPM

# Assuming pseq is your data frame or a similar object
col_without_confounders = setdiff(colnames(sample_data(pseq)), confounders)

# Iterate through each variable to run Maaslin2 and filter results
for(var in col_without_confounders) {
#for(var in c("ECC_status", "Child_dental_health")) { # for try
  # Run Maaslin2 analysis
  Maaslin2_res_Genus_ITS <- Maaslin2(input_data = t(otu_table(pseq)),
         input_metadata = sample_data(pseq) %>% data.frame(),
         output = Maaslin2_Genus_output_dir_ITS, # if file error quit sesssion and connect
         normalization = "CLR",
         transform = "NONE",
         fixed_effects =  c(var, confounders),
         cores = 40,
         max_significance = 1.1,
         min_prevalence = minimum_prevalence
           )
  
    # Filter and transform the output data frame
  Maaslin2_res_Genus_ITS_df <- Maaslin2_res_Genus_ITS$results %>% 
                 data.frame() %>% 
                  rename(OTUs = feature ) %>% 
                 filter(metadata  == var) %>% 
                 dplyr::select(OTUs, coef, qval, stderr)
  
  # Rename the coef and qval columns to the current variable
  df_coef_ITS <- Maaslin2_res_Genus_ITS_df %>% dplyr::select(OTUs, coef) %>% 
                     rename(!!var := coef)
                     
  df_qval_ITS <- Maaslin2_res_Genus_ITS_df %>% dplyr::select(OTUs, qval) %>% 
                      rename(!!var := qval)
  
  # Store in lists
  list_coef_ITS[[var]] <- df_coef_ITS
  list_qval_ITS[[var]] <- df_qval_ITS
}

df_coef_ITS <- list_coef_ITS[[1]]
# for(i in 2:length(list_coef_ITS)) {
#   Genus_coef_ITS <- full_join(Genus_coef_ITS, list_coef_ITS[[i]], by = "OTUs")
# }

df_qval_ITS <- list_qval_ITS[[1]]
# for(i in 2:length(list_qval_ITS)) {
#   Genus_qval_ITS <- full_join(Genus_qval_ITS, list_qval_ITS[[i]], by = "OTUs")
# }

Maaslin2_res_Genus_ITS
```

## combine 16S and ITS

```{r}
Maaslin2_res_Genus_16S_ITS = rbind(Maaslin2_res_Genus_16S_df, Maaslin2_res_Genus_ITS_df) %>% 
  # filter(qval < 0.01) %>%
  dplyr::select(OTUs, coef, qval, stderr)
Maaslin2_res_Genus_16S_ITS
```


## For DA Genus barplot
```{r}

Genus_coef_qval_16S_ITS = Maaslin2_res_Genus_16S_ITS %>% 
  rename("Coefficients" = coef) %>%
  rename("q_values" = qval) %>%
  rename("std_error" = stderr) %>%
  mutate(
    lower_ci = Coefficients - 1.96 * std_error,
    upper_ci = Coefficients + 1.96 * std_error
  ) %>% 
  # filter rows containing NA values
  filter(!is.na(Coefficients)) 
Genus_coef_qval_16S_ITS

saveRDS(Genus_coef_qval_16S_ITS,
         file = here("Samples538_A_Geno_Microbiome/A2_Maaslin",
                     "A2_05_Genus_coef_qval_16S_ITS.rds")
         )
```
```{r}
Genus_coef_qval_16S_ITS = readRDS(here("Samples538_A_Geno_Microbiome/A2_Maaslin",
                                       "A2_05_Genus_coef_qval_16S_ITS.rds")
                                 )
```


```{r}
# Data preparation
Genus_coef_qval_16S_ITS_2 <- Genus_coef_qval_16S_ITS %>% 
  filter(q_values < 0.01)
Genus_coef_qval_16S_ITS_2$log_q_values <- -log10(Genus_coef_qval_16S_ITS_2$q_values)
Genus_coef_qval_16S_ITS_2$Sign <- ifelse(Genus_coef_qval_16S_ITS_2$Coefficients > 0, "ECC", "CF")


# Use abbreviate to shorten names while avoiding duplicates
Genus_coef_qval_16S_ITS_2$Shortened_OTUs <- abbreviate(Genus_coef_qval_16S_ITS_2$OTUs, minlength = 35, use.classes = FALSE)
Genus_coef_qval_16S_ITS_2$Shortened_OTUs <- factor(Genus_coef_qval_16S_ITS_2$Shortened_OTUs, levels = Genus_coef_qval_16S_ITS_2$Shortened_OTUs[order(Genus_coef_qval_16S_ITS_2$Coefficients)])

Genus_coef_qval_16S_ITS_2
```


```{r}
Genus_coef_plot=
        ggplot(Genus_coef_qval_16S_ITS_2, aes(x = Shortened_OTUs, y = Coefficients, fill = Sign)) +
        ylab("Coefficient estimate (MaAsLin2)") +
        xlab("Microbial Genus") +
  scale_y_continuous(limits = c(-2, 2.5), breaks = seq(-2, 2, 1)) +
        theme_void() +
        theme(
            axis.title.y = element_blank(),
            axis.title.x = element_text(size = 11),
            axis.text.y  = element_blank(),
            axis.text.x  = element_text(vjust = 0.7, size = 9)) +
        geom_bar(
            stat = "identity",
            aes(fill = Sign),
            color = "black",
            size = 0.1,width = 0.75,
            alpha = 0.8) +
        theme(    # Legends
            legend.position = "top",
            legend.title = element_blank(),
            # legend.key.height = unit(0.07, 'cm'),
            # legend.key.width = unit(0.6, 'cm')
            ) +
        scale_fill_manual(values = c("ECC" = "darkred", "CF" = "darkblue")) +
        geom_text(    # Feature labeling
            aes(y = 0, label = Shortened_OTUs),
            hjust = ifelse(Genus_coef_qval_16S_ITS_2$Coefficients < 0, 0, 1),
            nudge_y = ifelse(Genus_coef_qval_16S_ITS_2$Coefficients < 0, 0.1, -0.1),
            color = "black",
            size = 2.5) +
        theme(    # Guide lines
            panel.grid.major.x = element_line(
                color = "grey", linewidth = 0.2, linetype = "dotted"),
            panel.grid.minor.x = element_line(
                color = "grey", linewidth = 0.1, linetype = "dotted")) +
        coord_flip()
Genus_coef_plot

# Save the plot
 ggsave(here("Samples538_A_Geno_Microbiome/A2_Maaslin",
             "A2_05_Maaslin2_Genus_barplot.jpg"),
       plot = Genus_coef_plot,
       width = 3, height = 2.7, dpi = 300)
 
 # Save the plot as rds
saveRDS(Genus_coef_plot,
         file = here("Samples538_A_Geno_Microbiome/A2_Maaslin",
                     "A2_05_Maaslin2_Genus_barplot.rds")
         )
 
```
